cli: lex sql input

madelynnblue opened the pull request:
Change the sql subcommand to lex the input so it can recognize semicolons
and \ commands only where expected.

Commands can now only be specified when there is not an active statement
being completed.

This requires tracking the current syntax mode. This is done by examining
the lex input for a SET token. If one is set, then the syntax is fetched
from the server after running the current group of statements.

Export sql/parser.scanner and add a helpful wrapper function to do the
lexing (since sqlSymType is still unexported).

The tests added are a very rudimentary start to #7334.

Fixes #4233

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7510)

<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[cli/sql.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNi4fPPMkz9hDE9GPC:-KLNi4fPPMkz9hDE9GPD:792887653) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/cli/sql.go#L327)):*
> ```Go
> 
> 		if hasSet {
> 			// TODO(mjibson): since the syntax is only updated once the commands are run,
> ```

On the server side, the syntax for parsing only changes between queries, not between statements in a multi-statement query. Isn't the current behavior here equivalent?

---

*[cli/sql_test.go, line 30 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNgX0_547SfqAvHwfy:-KLNgX0_547SfqAvHwfz:-1108242667) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/cli/sql_test.go#L30)):*
> ```Go
> //
> // CAUTION: this test cannot be copied and reused. I have not yet found a
> // convinent way to set readline.Stdin multiple times. Any subsequent use
> ```

`s/convinent/convenient/g`

---

*[cli/sql_test.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNgihiV3r0SN7BClSY:-KLNgihiV3r0SN7BClSZ:132391394) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/cli/sql_test.go#L33)):*
> ```Go
> // (i.e., if this test were copied) will result in a timeout due to hung go
> // routines. For now, we can have one giant input test string.
> // TODO(mjibson): enable multiple tests to use readline
> ```

Could we refactor `runInteractive` to make the source of input lines an interface instead of using `readline` directly?

I wonder if you should be testing `isEndOfStatement` directly. Doing so would allow greater coverage. For example, statements without a terminating semicolon, comment after the semicolon, etc.

---

*[sql/parser/scan.go, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNg3xco4FzqaQ3TyaD:-KLNg3xdvRfKr9af9ZHj:269724185) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/sql/parser/scan.go#L72)):*
> ```Go
> }
> 
> // Tokens calls f on tokens of the input until a non-zero token.
> ```

`s/on tokens/on all tokens/g`
`s/until a non-zero token/until an EOF is encountered/g`

---

*[sql/parser/scan.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNgIlwy5uhZcfsOf-z:-KLNgIlxiyPJz9J0grYh:-192187132) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/sql/parser/scan.go#L76)):*
> ```Go
> 	var sym sqlSymType
> 	for {
> 		t := s.Lex(&sym)
> ```

Should probably use `s.tokBuf` instead of `sym` here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-:-KLNienWAMdjHTjAs05B:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Are there any error cases for which the behavior changes? Currently if we send an invalid statement in a transaction, the transaction gets aborted. We don't want that behavior to depend on whether the lexer catches the error or not. I assume nothing changes since we still send everything to the server, but wanted to doublecheck.

mjibson commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[cli/sql.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNi4fPPMkz9hDE9GPC:-KLNm4w21nXUlBgMB9aa:-2053471244) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/cli/sql.go#L327)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> On the server side, the syntax for parsing only changes between queries, not between statements in a multi-statement query. Isn't the current behavior here equivalent?
</details>
I wasn't aware of that. Removed this TODO.

---

*[cli/sql_test.go, line 30 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNgX0_547SfqAvHwfy:-KLNmMpLCFaXUBbb2AqX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/cli/sql_test.go#L30)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/convinent/convenient/g`
</details>
Done.

---

*[cli/sql_test.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNgihiV3r0SN7BClSY:-KLNnDsuUAWZAF-PipiY:940387367) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/cli/sql_test.go#L33)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Could we refactor `runInteractive` to make the source of input lines an interface instead of using `readline` directly?
> 
> I wonder if you should be testing `isEndOfStatement` directly. Doing so would allow greater coverage. For example, statements without a terminating semicolon, comment after the semicolon, etc.
</details>
I've refactored `runInteractive` to take a readline Config and added tests for isEndOfStatement.

---

*[sql/parser/scan.go, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNg3xco4FzqaQ3TyaD:-KLNm_2FEvwD318irGV5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/sql/parser/scan.go#L72)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/on tokens/on all tokens/g`
> `s/until a non-zero token/until an EOF is encountered/g`
</details>
Done.

---

*[sql/parser/scan.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-KLNgIlwy5uhZcfsOf-z:-KLNn0coKmxIBlWX_46l:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f25ecb856a1333fd03508aa17312b3f292c09a06/sql/parser/scan.go#L76)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Should probably use `s.tokBuf` instead of `sym` here. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-:-KLT2aGV5Rr8tgoEKCFL:1417088721)*
<!-- Sent from Reviewable.io -->

mjibson commented:
I can't think of a new error case that this change introduces, so I think we're safe.

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7510#-:-KLTBaZFGSKPnDQBk9I6:-1694397706)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
cli: lex sql input

Change the sql subcommand to lex the input so it can recognize semicolons
and \ commands only where expected.

Commands can now only be specified when there is not an active statement
being completed.

This requires tracking the current syntax mode. This is done by examining
the lex input for a SET token. If one is set, then the syntax is fetched
from the server after running the current group of statements.

Export sql/parser.scanner and add a helpful wrapper function to do the
lexing (since sqlSymType is still unexported).

The tests added are a very rudimentary start to #7334.

Fixes #4233
